deepfence_agent/tools/apache/scope/probe/host/secret_scanner.go

```diff
@@ -0,0 +1,216 @@
+package host
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/weaveworks/scope/common/xfer"
+	"github.com/weaveworks/scope/probe/secret_scanner"
+	"google.golang.org/grpc"
+	"github.com/gomodule/redigo/redis"
+	elastic "github.com/olivere/elastic/v7"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	secretScanIndexName     = "secret-scan"
+	secretScanLogsIndexName = "secret-scan-logs"
+	redisSecretScanChannel  = "secret_scan_task_queue"
+)
+var certPath = "/etc/filebeat/filebeat.crt"
+var httpClient *http.Client
+
+func (r *Reporter) startSecretsScan(req xfer.Request) xfer.Response {
+	nodeType := fmt.Sprintf("%s", req.ControlArgs["node_type"])
+	var greq secret_scanner.FindRequest
+	if nodeType == nodeTypeContainer {
+		containerID := fmt.Sprintf("%s", req.ControlArgs["container_id"])
+		if containerID == "" {
+			return xfer.ResponseErrorf("container_id is required")
+		}
+		greq = secret_scanner.FindRequest{Input: secret_scanner.FindRequest_Container{
+			Container: secret_scanner.Container{Id: containerID},
+		}}
+	} else if nodeType == nodeTypeImage {
+		imageId := fmt.Sprintf("%s", req.ControlArgs["image_id"])
+		if imageId == "" {
+			return xfer.ResponseErrorf("image_id is required")
+		}
+		imageName := fmt.Sprintf("%s", req.ControlArgs["image_id"])
+		greq = secret_scanner.FindRequest{Input: secret_scanner.FindRequest_Image{
+			Image: secret_scanner.DockerImage{Id: imageId, Name: imageName},
+		}}
+	} else if nodeType == nodeTypeHost {
+		greq = secret_scanner.FindRequest{Input: secret_scanner.FindRequest_Path{Path: HostMountDir}}
+	}
+	client, err := newSecretScannerClient()
+	if err != nil {
+		return xfer.Response{SecretsScanInfo: "Error creating ss client"}
+	}
+	go getAndPublishSecretScanResults(client, greq)
+	return xfer.Response{SecretsScanInfo: "Secrets scan started"}
+}
+
+func getAndPublishSecretScanResults(client secret_scanner.SecretScannerClient, req secret_scanner.FindRequest) {
+	// TODO get scan id from df-backend
+	res, err := client.FindSecretInfo(context.Background(), req)
+	if err != nil {
+		scanLog := fmt.Sprintf("{\"scan_id\":\"%d\",\"time_stamp\":%d,\"secret_scan_message\":\"%s\",\"action\":\"%s\",\"type\":\"secret-scan\",}", getTimestamp(), getTimestamp(), err.Error(), "ERROR")
+		sendSecretScanDataToLogstash(scanLog, secretScanLogsIndexName)
+	}
+	for _, secret := range res.Secrets {
+		// TODO handle marshal error
+		secretString, _ := json.Marshal(secret)
+		scanLog := fmt.Sprintf("{\"scan_id\":\"%d\",\"time_stamp\":%d,\"secret_scan_data\":\"%s\"}", getTimestamp(), getTimestamp(), string(secretString))
```

Comment:
Code scanning

Potentially unsafe quoting
Critical

If this JSON value contains a double quote, it could break out of the enclosing quotes.